### ISSUE TITLE
Shadowing of `@last`/`@first` by ListOrSingleton#getLast() in Response Templates

### ISSUE THREAD

**Issue by LitschiW** [wiremock/wiremock#2204]
### TLDR
Greetings,
In the past couple of days, I ran into a problem getting the following response template to work:
(It simply replies all query parameters as json.)
```json 
"[{\n {{#each request.query}} \"{{@key}}\": \"{{this}}\"{{#unless @last}},{{/unless}}\n{{/each}} }]"
```
The comma from the `#unless` function was always missing. It would be expected to appear in all but the last iteration. And in fact, inside the `#each` function, the `@last` data variable itself was not working correctly and always contained the same value as the `@key`, but should be empty. So, I dug into the code, figured out the problem, wrote a [quick and dirty solution](https://gist.github.com/LitschiW/f088410458e774657e138f141baf0645) and figured I'll share my experiences here for others to find (and maybe point out an area of improvement for Wiremock 3 ;) ). 

---
### Further explanation
First of, I am using wiremock 2.27.2 with handlebars.java 4.31.2.

Response templates are parsed using [handelbars.java](https://github.com/jknack/handlebars.java). 
When trying to use [`@first` or `@last`](https://handlebarsjs.com/api-reference/data-variables.html#first) inside an `#each` function, that iterates over the query parameters, these @data variables contain the values of the first or last elements of its wrapping `ListOrSingle` object. Instead, they should contain "first" or "last" on the first and last iteration, respectively.

And here's why:  The underlying [`JavaBeanValueResolver`](https://github.com/jknack/handlebars.java/blob/master/handlebars/src/main/java/com/github/jknack/handlebars/context/JavaBeanValueResolver.java) tries to match the variable name (e.g. "last") against the member names of the data object (`ListOrSingle`). Since `ListOrSingle` implements both `getFirst` and `getLast` (as forced part of its inheritance)  the resolver uses these getter methods to retrieve the, most often truthy, values. However, it would not be expected to find any values at all. This obscures the actual value set to `@last` and `@first` causing them to almost always be truthy. Hence, the `#unless` function is never applied.

The rather [simple solution I came up with](https://gist.github.com/LitschiW/f088410458e774657e138f141baf0645) is to build a Helper/new function `#first` that digs into the extended context that is given by handlebars using reflection. This is neither a pretty nor very stable solution (e.g., it does not support `PathExpressions`) but it works for my simple circumstances. 
This helper obviously has to be registered to the handlebars instance. For this, need to create a [custom `WireMockWebContextListener`](https://gist.github.com/LitschiW/68fa1bd07c20de8da9c5383010f9bf1d), that is basically a copy, but uses a `WireMockConfiguration` instead of a `WarConfiguration`.

Two alternative long-term solutions I thought about during my progress:
- Let `ListOrSingle` inherit from `AbstractList` instead of `ArrayList`. Which might even makes sense semantically if those values are usually immutable(?). However, this is rather breaking.
-  Add new Helpers/utility functions, such as mine, to your `Handlebars` instance and mark `@last`/`@first` as potentially broken in the `#each` context.

### Reproduction steps

Test for reproduction:
```java 
import com.github.jknack.handlebars.Handlebars;
import com.github.tomakehurst.wiremock.common.ListOrSingle;
import com.github.tomakehurst.wiremock.extension.responsetemplating.HandlebarsOptimizedTemplate;
import org.junit.Assert;
import org.junit.Test;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class LastInEachTest {

    @Test
    public void testLastInEach() throws IOException {
        final var expected =
                ("[{ " +
                 "`a`: `b`,\n" +
                 "`c`: `[d]`,\n" +
                 "`e`: `[f]`,\n" +
                 "`g`: `h` " +
                 "}]").replaceAll("`", "\"").replaceAll("\\s+", "");

        final var templateString = "[{\n {{#each request.query}} \"{{@key}}\": \"{{this}}\"{{#unless @last}},{{/unless}}\n{{/each}} }]";
        final var handlebars = new Handlebars();

        final var template = new HandlebarsOptimizedTemplate(handlebars, templateString);

        Map<String, List<String>> tmp = Map.of("a", new ListOrSingle<>("b"),
                                               "c", List.of("d"),
                                               "e", new LinkedList<>() {{add("f");}},
                                               "g", new ListOrSingle<>("h"));

        var result = template.apply(Map.of("request", Map.of("query", new TreeMap<>(tmp))));

        Assert.assertEquals(expected, result.replaceAll("\\s+", ""));
    }
}
```
produces: 
```
Expected : [{"a":"b","c":"[d]","e":"[f]","g":"h"}]
Actual   : [{"a":"b""c":"[d]","e":"[f]""g":"h"}]
```

Notice, that `List#of` produces a List that does not implement `getLast` (`AbstractImmutableList`), hence the comma appears correctly in the middle of the output.

### References

- [FirstAndLastHelpers.java](https://gist.github.com/LitschiW/f088410458e774657e138f141baf0645)
- [CustomWireMockWebContextListener.java](https://gist.github.com/LitschiW/68fa1bd07c20de8da9c5383010f9bf1d)

If you want to see the wrong resolve yourself, you can put a conditional break point  (`head.toString().equals("@last")`) in [line 618 of Context.java](https://github.com/jknack/handlebars.java/blob/1ca4f1ba22680cc313e23ca709f24decb9990316/handlebars/src/main/java/com/github/jknack/handlebars/Context.java#L618) of handlebars.java, debug the test above and dive into the application of the resolvers. Eventually, you will end up executing the [JavaBeanValueResolver.java](https://github.com/jknack/handlebars.java/blob/master/handlebars/src/main/java/com/github/jknack/handlebars/context/JavaBeanValueResolver.java) which has a cache, that shows that "last" is mapped to `ListOrSingle#getLast`.
